bin/i18n/translation_monitor/crowdin_validator.rb

```diff
@@ -0,0 +1,348 @@
+
+    # Validate markdown content
+    error_messages << validate_markdown_link(translation.text)
+
+    # Validate translation language
+    languages = load_languages
+    accepted_languages = languages[translation.crowdin_language_id]['language_detector_code'] || translation.crowdin_language_id
+    error_messages << validate_language(translation.text, accepted_languages)
+
+    # Combine error messages found by the validation functions with
+    # other useful information to help translators fixing errors faster.
+    crowdin_editor_code = languages[translation.crowdin_language_id]['editor_code']
+    project_name = Crowdin::Client::CDO_PROJECT_IDS.key(source.project_id)
+    project_source_language = Crowdin::Client::CDO_PROJECT_SOURCE_LANGUAGES[project_name]
+    crowdin_link = "https://crowdin.com/translate/#{project_name}/#{source.file_id}/#{project_source_language}-#{crowdin_editor_code}##{source.id}"
+    common_info = {
+      string_id: source.id,
+      string: source.text,
+      string_created_at: source.created_at,
+      translation_id: translation.id,
+      translation: translation.text,
+      translator_id: translation.user_id,
+      translator_user_name: translation.user_name,
+      translation_created_at: translation.created_at,
+      crowdin_link: crowdin_link
+    }
+
+    [].tap do |errors|
+      error_messages.each do |error_message|
+        next if error_message.nil? || error_message.empty?
+        errors << common_info.merge({error_message: error_message}) unless error_message.empty?
+      end
+    end
+  end
+
+  # Transform an array of translation to a hash for faster lookup.
+  # @param translations [Array<Hash>]
+  # @return [Hash{string_id => Hash{translation_id => Translation}}]
+  def transform_translations(translations)
+    {}.tap do |res|
+      translations.each do |data|
+        string_id = data['stringId']
+        translation_id = data['translationId']
+        user_id = data.dig('user', 'id')
+        raise 'string_id must not be nil' if string_id.nil?
+        raise 'translation_id must not be nil' if translation_id.nil?
+        raise 'user_id must not be nil' if user_id.nil?
+
+        res[string_id] ||= {}
+        res[string_id][translation_id] = Translation.new(
+          id: translation_id,
+          string_id: string_id,
+          text: data['text'],
+          crowdin_language_id: data['crowdin_language_id'],
+          user_id: user_id,
+          user_name: data.dig('user', 'username'),
+          created_at: data['createdAt']
+        )
+      end
+    end
+  end
+
+  # Transform an array of source strings to a hash for faster lookup.
+  # @param source_strings [Array<Hash>]
+  # @return [Hash{string_id => SourceString}]
+  def transform_source_strings(source_strings)
+    {}.tap do |res|
+      source_strings.each do |data|
+        string_id = data['id']
+        raise 'string_id must not be nil' if string_id.nil?
```

Comment:
same comment as above.

Reply:
will remove, this was only for development